single-pool/js/src/transactions.ts

```diff
@@ -0,0 +1,302 @@
+      }),
+    );
+
+    transaction.add(
+      SystemProgram.transfer({
+        fromPubkey: payer,
+        toPubkey: stake,
+        lamports: stakeRent + minimumDelegation,
+      }),
+    );
+
+    transaction.add(
+      SystemProgram.transfer({
+        fromPubkey: payer,
+        toPubkey: mint,
+        lamports: mintRent,
+      }),
+    );
+
+    transaction.add(SinglePoolInstruction.initializePool(voteAccount));
+
+    if (!skipMetadata) {
+      transaction.add(SinglePoolInstruction.createTokenMetadata(pool, payer));
+    }
+
+    return transaction;
+  }
+
+  static async deposit(params: DepositParams) {
+    const { connection, pool, userWallet } = params;
+
+    if (!params.userStakeAccount == !params.depositFromDefaultAccount) {
+      throw 'must either provide userStakeAccount or true depositFromDefaultAccount';
+    }
+
+    const userStakeAccount = params.depositFromDefaultAccount
+      ? await findDefaultDepositAccountAddress(pool, userWallet)
+      : params.userStakeAccount;
+
+    const transaction = new Transaction();
+
+    const mint = findPoolMintAddress(this.programId, pool);
+    const poolStakeAuthority = findPoolStakeAuthorityAddress(this.programId, pool);
+    const userAssociatedTokenAccount = getAssociatedTokenAddressSync(mint, userWallet);
+
+    const userTokenAccount = params.userTokenAccount || userAssociatedTokenAccount;
+    const userLamportAccount = params.userLamportAccount || userWallet;
+    const userWithdrawAuthority = params.userWithdrawAuthority || userWallet;
+
+    if (
+      userTokenAccount.equals(userAssociatedTokenAccount) &&
+      (await connection.getAccountInfo(userAssociatedTokenAccount)) == null
+    ) {
+      transaction.add(
+        createAssociatedTokenAccountInstruction(
+          userWallet,
+          userAssociatedTokenAccount,
+          userWallet,
+          mint,
+        ),
+      );
+    }
+
+    transaction.add(
+      StakeProgram.authorize({
+        stakePubkey: userStakeAccount as PublicKey,
+        authorizedPubkey: userWithdrawAuthority,
+        newAuthorizedPubkey: poolStakeAuthority,
+        stakeAuthorizationType: StakeAuthorizationLayout.Staker,
+      }),
+    );
+
+    transaction.add(
+      StakeProgram.authorize({
+        stakePubkey: userStakeAccount as PublicKey,
+        authorizedPubkey: userWithdrawAuthority,
+        newAuthorizedPubkey: poolStakeAuthority,
+        stakeAuthorizationType: StakeAuthorizationLayout.Withdrawer,
+      }),
+    );
+
+    transaction.add(
+      SinglePoolInstruction.depositStake(
+        pool,
+        userStakeAccount as PublicKey,
+        userTokenAccount,
+        userLamportAccount,
+      ),
+    );
+
+    return transaction;
+  }
+
+  static async withdraw(params: WithdrawParams) {
+    const { connection, pool, userWallet, userStakeAccount, tokenAmount, createStakeAccount } =
+      params;
+
+    if (typeof tokenAmount == 'bigint' && tokenAmount > BigInt(Number.MAX_SAFE_INTEGER)) {
```

Comment:
Shouldn't this be:
```suggestion
    if (typeof tokenAmount != 'bigint' && tokenAmount > BigInt(Number.MAX_SAFE_INTEGER)) {
```

If it's a bigint, it should be safe if it's big, right?

Reply:
no, we have to convert to number when passing to web3 because thats what it supports, so this check is "if its a bigint, error if that conversion would be lossy"

Reply:
`tokenAmount` is passed to your own instruction, which supports `bigint`, and to `createApproveInstruction`, which also supports `bigint`, so there's no raw web3.js stuff to worry about from what I can see. Is there something else that I'm missing?

Reply:
oh yes youre right, spl-token supports bigint. web3.js doesnt, and i had made this "fix" to `withdraw` after discovering i needed to convert to number in `createAndDelegateUserStake`